tests/hardhat/EvilXToken.ts

```diff
@@ -0,0 +1,205 @@
+import { convertToUnit } from "../../helpers/utils";
+import chai from "chai";
+import { ethers } from "hardhat";
+import { smock } from "@defi-wonderland/smock";
+
+import { VAIControllerHarness__factory, ComptrollerHarness__factory, IAccessControlManager } from "../../typechain";
+
+const { expect } = chai;
+
+describe("Evil Token test", async () => {
+  let vToken1, vToken2, vToken3, unitroller;
+  it("Check the updated vToken states after transfer out", async () => {
+    const [root, account1] = await ethers.getSigners();
+
+    const accessControlMock = await smock.fake<IAccessControlManager>("AccessControlManager");
+    accessControlMock.isAllowedToCall.returns(true);
+
+    const user = account1;
+    const cf1 = 0.5,
+      cf2 = 0.666,
+      cf3 = 0,
+      up1 = 3,
+      up2 = 2.718,
+      up3 = 1;
+
+    const comptrollerLensFactory = await ethers.getContractFactory("ComptrollerLens");
```

Comment:
We should have deployment scripts that deploy contracts for tests instead of manually deploying them in the test.

Hardhat-deploy will do that when you run `hardhat test`. I am curious though how you would run tests withouy having state dependent on other tests

Reply:
1. This is not in the scope of this ticket. 
2. The deployment script for the core pool has not been made till now thats why we are using hardhat upgrade to deploy the tests.
3. Upgradation of these contracts will be done manually.
cc: @kkirka

Reply:
I agree refactoring to use deploy scripts is out of scope of this ticket. What brought this to my attention is the really long function. This is a code smell and can be difficult to maintain. In reading through our tests I notice lots of duplicate code. Writing new tests so that they are more maintainable should be our goal. 

The easiest to improve this for now would be to break it out into functions that can be used in this test and others.

Another idea on how we could improve this is to move all the deployments that we expect `hardhat-deploy` to replace into a separate function. That way when we do write the deployment scripts it is clear how to refactor this test.

Reply:
or an even simpler refactoring that improves on the way this test is currently written would be to remove setup (deployments) to a before hook

Reply:
Yes, we should use deployments in the before hook if tests need to deploy the same contracts several time or for multiple tests, but here contracts are deployed one time only thats why I did not use before hook.

Reply:
I think the main reason for the argument here is that this test looks a lot like an integrational test: it deploys multiple contracts and only makes use of a limited number of fakes. This resembles the old "scenario" tests.

I agree that such complex tests should use deployment scripts. I don't have any opinion wrt this ticket vs other ticket, I'm fine with either.

I don't think unit tests should use the deployment scripts though. In unit tests we're interested in the behavior of a particular contract in isolation: whether it indeed makes external calls, whether the state of this contract is as expected, whether this contract emits the necessary events, etc. For such tests, we'd need a specific deployment – usually the tested contract would be a `smock.mock<...>`, and most of the other contracts would be `smock.fake<...>` (with an exception of ERC-20 tokens and maybe comptroller lens when testing comptroller as it is an irrevokable part of the comptroller itself).

Reply:
yeah that makes a lot of sense that unit tests would use the hardhat deploy scripts. I still think we should have a pattern that reduces the size of test cases so that they are easier to read and maintain. I think generally if different tests have different deployments they would be in different suites so the before hook is a safe place to deploy contracts

Reply:
ok will add it.

Reply:
added.